Detect escaped quotation marks in readQuotedString

mathisgerdes opened the pull request:
Within quoted string quotation marks may appear as `...\"...`, which should not terminate the quote.
I observed this multiple times in fetch responses from gmail, leading to an error in the previous version.

mathisgerdes committed:
Detect escaped quotation marks in readQuotedString

Within quoted string quotation marks may appear as ...\"... which
should not terminate the quote.

mathisgerdes commented:
According to [RFC 2683](https://tools.ietf.org/html/rfc2683#section-3.4.2) the two cases to be handled are `\"` and `\\`.

mathisgerdes committed:
Fix typo and handle all escape cases in RFC 2683

mathisgerdes commented:
I am not sure how you would like to handle the case of `\` followed by something other than `\` or `"`. Currently the chars are added verbatim (i.e. treating `\<not \ or ">` as `\\<not \ or ">`); I don't know if you'd rather throw an error.

michaelspiss commented:
Thank you for reporting this, I cannot remember anything like that from reading the rfc. Could you please open an issue for this so we can discuss it over there (preferably with logs)? I am currently on vacation, so I won't be able to confirm anything until I get back home, but we should at least be able to (1) get the logs ready for me to reproduce the error and (2) maybe already provide a fix that only needs to be merged

mathisgerdes commented:
Please excuse my potential confusion at this point, but is this pull request not a way to provide a fix that only needs to be merged? I'm not quite sure which logs you refer to, besides maybe the explicit error which occurs (the code continues to scan part of the subject line and fails with an unexpected imap word). 

To reproduce the error try to fetch the envelop of an email that contains a quotation mark in the subject line.

As mentioned in a previous comment, you can read about the escaped characters here:
[https://tools.ietf.org/html/rfc2683#section-3.4.2](https://tools.ietf.org/html/rfc2683#section-3.4.2)

mathisgerdes committed:
Reformat Code.

michaelspiss commented:
I am so sorry for the confusion, I missed the comment where you mentioned rfc 2683 on my mobile - thus the strange response. With logs I meant this package's logs which can be viewed by calling printImapClientDebugLog(), but this is obsolete due to the nature of this issue.

The only thing missing would be to add one or two tests for this case so it won't be overlooked in future versions (in test/imap_buffer_test.dart), I will merge this pull request as soon as I'm back home. 

Again, I am terribly sorry for the confusion caused.

michaelspiss commented:
After a lot of consideration I think it's best to throw a SyntaxErrorException if the character to escape is neither a backslash nor a quotation mark. Whilst every character is escapable, most simply don't have any meaning to the protocol, plus some other characters have special meanings when escaped (\t for tab, etc.) which would produce wrong results.

mathisgerdes committed:
Add tests and throw SyntaxErrorException

Add tests for escaped characters in quoted string and throw
a SyntaxErrorException if an escape sequence besides backslash or
quotation mark appears.

michaelspiss committed:
Remove quotes and escape characters from comments

This is to avoid confusion regarding the new escaped characters and single - non-escaped - characters

michaelspiss commented:
Thank you for your help!
